azure-ipam/ipam_test.go

```diff
@@ -0,0 +1,283 @@
+		Name:       "happynetconf",
+	}
+
+	invalidVersionNetConf := &cniTypes.NetConf{
+		CNIVersion: "0",
+		Name:       "nothappynetconf",
+	}
+
+	happyNetConfByteArr, err := json.Marshal(happyNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidVersionNetConfByteArr, err := json.Marshal(invalidVersionNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidNetConf := []byte("invalidNetConf")
+
+	tests := []scenario{
+		{
+			name: "Happy CNI add",
+			args: buildArgs("happyArgs", happyPodArgs, happyNetConfByteArr),
+			want: &types100.Result{
+				CNIVersion: "1.0.0",
+				IPs: []*types100.IPConfig{
+					{
+						Address: net.IPNet{
+							IP:   net.IPv4(10, 0, 1, 10),
+							Mask: net.CIDRMask(24, 32),
+						},
+						Gateway: net.IPv4(10, 0, 0, 1),
+					},
+				},
+				Routes: []*cniTypes.Route{
+					{
+						Dst: net.IPNet{
+							IP:   net.IPv4(0, 0, 0, 0),
+							Mask: net.CIDRMask(0, 32),
+						},
+						GW: net.IPv4(10, 0, 0, 1),
+					},
+				},
+				DNS: cniTypes.DNS{},
+			},
+			wantErr: false,
+		},
+		{
+			name:    "Fail request CNS ipconfig during CmdAdd",
+			args:    buildArgs("failRequestCNSArgs", happyPodArgs, happyNetConfByteArr),
+			wantErr: true,
+		},
+		{
+			name:    "Fail process CNS response during CmdAdd",
+			args:    buildArgs("failProcessCNSResp", happyPodArgs, happyNetConfByteArr),
+			wantErr: true,
+		},
+		{
+			name:    "Fail parse netconf during CmdAdd",
+			args:    buildArgs("failParseNetConf", happyPodArgs, invalidNetConf),
+			wantErr: true,
+		},
+		{
+			name:    "Fail get versioned result during CmdAdd",
+			args:    buildArgs("failGetVersionedResult", happyPodArgs, invalidVersionNetConfByteArr),
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		writer := &cniResultsWriter{}
+		t.Run(tt.name, func(t *testing.T) {
+			fmt.Printf("test : %v\n", tt.name)
+			mockCNSClient := &MockCNSClient{}
+			testLogger, cleanup, err := logger.New(loggerCfg)
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+			defer cleanup()
+			ipamPlugin, _ := NewPlugin(testLogger, mockCNSClient, writer)
+			err = ipamPlugin.CmdAdd(tt.args)
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+			if tt.want != nil {
+				require.Equal(t, tt.want, writer.result)
+			}
+		})
+	}
+}
+
+func TestCmdDel(t *testing.T) {
+	happyNetConf := &cniTypes.NetConf{
+		CNIVersion: "1.0.0",
+		Name:       "happynetconf",
+	}
+
+	happyNetConfByteArr, err := json.Marshal(happyNetConf)
+	if err != nil {
+		panic(err)
+	}
+
+	tests := []scenario{
+		{
+			name:    "Happy CNI del",
+			args:    buildArgs("happyArgs", happyPodArgs, happyNetConfByteArr),
+			wantErr: false,
+		},
+		{
+			name:    "Fail request CNS release IP during CmdDel",
+			args:    buildArgs("failRequestCNSReleaseIPArgs", happyPodArgs, happyNetConfByteArr),
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			fmt.Printf("test : %v\n", tt.name)
```

Comment:
nit: the Go test framework should print the (sub)test name for you, I think? (Same for the other tests in this file)

Reply:
@nddq tests shouldn't print directly to stdout, they should just pass or fail with message - `testify`'s assert and require functions have an optional message parameter at the end. Parallel tests all printing to stdout just make a mess 🥲

Reply:
should I try using Ginko 🙂